src/services/refactors/moveToNewFile.ts

```diff
@@ -0,0 +1,359 @@
+/* @internal */
+namespace ts.refactor {
+    const refactorName = "Move to new file";
+    registerRefactor(refactorName, {
+        getAvailableActions(context): ApplicableRefactorInfo[] {
+            if (getStatementsToMove(context) === undefined) return undefined;
+            const description = getLocaleSpecificMessage(Diagnostics.Move_to_new_file);
+            return [{ name: refactorName, description, actions: [{ name: refactorName, description }] }];
+        },
+        getEditsForAction(context, actionName): RefactorEditInfo {
+            Debug.assert(actionName === refactorName);
+            const statements = Debug.assertDefined(getStatementsToMove(context));
+            const edits = textChanges.ChangeTracker.with(context, t => doChange(context.file, context.program, statements, t, context.host));
+            return { edits, renameFilename: undefined, renameLocation: undefined };
+        }
+    });
+
+    function getStatementsToMove(context: RefactorContext): ReadonlyArray<Statement> | undefined {
+        const { file } = context;
+        const range = createTextRangeFromSpan(getRefactorContextSpan(context));
+        const { statements } = file;
+
+        const startNodeIndex = findIndex(statements, s => s.end > range.pos);

```

Comment:
so that means that if i select a nested function to move, we will move the containing function instead.. would not it be better to not make the refactoring available at this case?

Reply:
See `moveToNewFile_rangeInvalid.ts`.